File: phases/03-data-structures/08-binary-search-trees-rotations/code/main.py
"""main.py — BST + rotations in Python.

The main point is to see in code that LEFT rotation preserves inorder.
"""
from __future__ import annotations
from dataclasses import dataclass
import random


@dataclass
class Node:
    key: int
    left: "Node | None" = None
    right: "Node | None" = None


def insert(n: Node | None, k: int) -> Node:
    if n is None: return Node(k)
    if k < n.key: n.left  = insert(n.left, k)
    elif k > n.key: n.right = insert(n.right, k)
    return n


def contains(n: Node | None, k: int) -> bool:
    while n:
        if k < n.key: n = n.left
        elif k > n.key: n = n.right
        else: return True
    return False


def height(n: Node | None) -> int:
    if n is None: return 0
    return 1 + max(height(n.left), height(n.right))


def rotate_left(n: Node) -> Node:
    r = n.right
    assert r is not None
    n.right = r.left
    r.left = n
    return r


def rotate_right(n: Node) -> Node:
    l = n.left
    assert l is not None
    n.left = l.right
    l.right = n
    return l


def inorder(n: Node | None) -> list[int]:
    out: list[int] = []
    def go(n):
        if n is None: return
        go(n.left); out.append(n.key); go(n.right)
    go(n); return out


def main() -> None:
    bad = None
    for i in range(1, 1001):
        bad = insert(bad, i)
    print(f"sorted insert n=1000 → height = {height(bad)} (expected 1000)")

    random.seed(42)
    good = None
    for _ in range(1000):
        good = insert(good, random.randrange(100000))
    print(f"random insert n=1000 → height = {height(good)} (expected ~20)")

    t = None
    for v in (10, 5, 20, 15, 25): t = insert(t, v)
    pre = inorder(t)
    t = rotate_left(t)
    post = inorder(t)
    print(f"\nbefore rotate_left: {pre}")
    print(f"after  rotate_left: {post}   ← inorder unchanged: invariant preserved")
    print(f"new root: {t.key}")


if __name__ == "__main__":
    main()

File: phases/03-data-structures/08-binary-search-trees-rotations/code/main.rs
//! main.rs — BST + rotations in safe Rust.

pub struct Node {
    key: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn new_node(k: i32) -> Box<Node> { Box::new(Node { key: k, left: None, right: None }) }

pub fn insert(n: Option<Box<Node>>, k: i32) -> Option<Box<Node>> {
    match n {
        None => Some(new_node(k)),
        Some(mut b) => {
            if k < b.key      { b.left  = insert(b.left.take(),  k); }
            else if k > b.key { b.right = insert(b.right.take(), k); }
            Some(b)
        }
    }
}

pub fn contains(mut n: Option<&Node>, k: i32) -> bool {
    while let Some(c) = n {
        if k < c.key      { n = c.left.as_deref(); }
        else if k > c.key { n = c.right.as_deref(); }
        else { return true; }
    }
    false
}

pub fn height(n: &Option<Box<Node>>) -> i32 {
    match n {
        None => 0,
        Some(b) => 1 + height(&b.left).max(height(&b.right)),
    }
}

pub fn rotate_left(mut n: Box<Node>) -> Box<Node> {
    let mut r = n.right.take().expect("right must exist for rotate_left");
    n.right = r.left.take();
    r.left = Some(n);
    r
}

pub fn rotate_right(mut n: Box<Node>) -> Box<Node> {
    let mut l = n.left.take().expect("left must exist for rotate_right");
    n.left = l.right.take();
    l.right = Some(n);
    l
}

pub fn inorder(n: &Option<Box<Node>>, out: &mut Vec<i32>) {
    if let Some(b) = n {
        inorder(&b.left, out);
        out.push(b.key);
        inorder(&b.right, out);
    }
}

fn main() {
    let mut t: Option<Box<Node>> = None;
    for v in [10, 5, 20, 15, 25] { t = insert(t, v); }

    let mut pre = vec![]; inorder(&t, &mut pre);
    println!("inorder before rotate_left: {:?}", pre);

    let t = rotate_left(t.unwrap());
    let mut post = vec![]; inorder(&Some(t), &mut post);
    println!("inorder after  rotate_left: {:?}   (unchanged: invariant preserved)", post);
}
